crates/datasources/src/bson/errors.rs

```diff
@@ -1,8 +1,14 @@
+use datafusion::arrow::datatypes::DataType;
 use datafusion::error::DataFusionError;
 use datafusion_ext::errors::ExtensionError;
 
 use crate::object_store::errors::ObjectStoreSourceError;
 
+/// Recursion limit for inferring the schema for nested documents.
+///
+/// The MongoDB kernel rejects nesting of greater than 100.
+pub const RECURSION_LIMIT: usize = 100;
```

Comment:
I don't know if this has always been 100, but do we want to match mongo's nesting limit here? This _could_ potentially be very expensive on our side, all for just inferring the schema (I also have no data to back up how fast schema inference is).

Reply:
I have a fair amount of confidence in the rust bson library from a performance perspective, so I'm not super worried about that code (and the message serialization work is probably more expensive than our work here.) 

I think if we had more strict limits than MongoDB that could lead us to reject data in ways that would be more expensive, and that feels like it could be bad for users. 

MongoDB has rejected documents with more than 100 levels since 2013 or so, and I'm not aware of any real complaints. about it, so I think we don't need a larger limit. 100 Is sort of arbitrary, not _actually_ that big and mostly prevents us from blowing up the stack or letting a user bomb the stack.

Reply:
Sounds good, we can stick with 100 here.

Reply:
(I just duplicated this constant here, so it's consistent with the existing behavior)